File: ClassEval_T/java/solutuon/MetricsCalculator.java
package org.example;

public class MetricsCalculator {
    int truePositives;
    int falsePositives;
    int falseNegatives;
    int trueNegatives;

    public MetricsCalculator() {
        this.truePositives = 0;
        this.falsePositives = 0;
        this.falseNegatives = 0;
        this.trueNegatives = 0;
    }

    public void update(int[] predictedLabels, int[] trueLabels) {
        for (int i = 0; i < predictedLabels.length; i++) {
            if (predictedLabels[i] == 1 && trueLabels[i] == 1) {
                this.truePositives++;
            } else if (predictedLabels[i] == 1 && trueLabels[i] == 0) {
                this.falsePositives++;
            } else if (predictedLabels[i] == 0 && trueLabels[i] == 1) {
                this.falseNegatives++;
            } else if (predictedLabels[i] == 0 && trueLabels[i] == 0) {
                this.trueNegatives++;
            }
        }
    }

    public double precision(int[] predictedLabels, int[] trueLabels) {
        update(predictedLabels, trueLabels);
        if (this.truePositives + this.falsePositives == 0) {
            return 0.0;
        }
        return (double) this.truePositives / (this.truePositives + this.falsePositives);
    }

    public double recall(int[] predictedLabels, int[] trueLabels) {
        update(predictedLabels, trueLabels);
        if (this.truePositives + this.falseNegatives == 0) {
            return 0.0;
        }
        return (double) this.truePositives / (this.truePositives + this.falseNegatives);
    }

    public double f1Score(int[] predictedLabels, int[] trueLabels) {
        update(predictedLabels, trueLabels);
        double precision = precision(predictedLabels, trueLabels);
        double recall = recall(predictedLabels, trueLabels);
        if (precision + recall == 0.0) {
            return 0.0;
        }
        return (2 * precision * recall) / (precision + recall);
    }

    public double accuracy(int[] predictedLabels, int[] trueLabels) {
        update(predictedLabels, trueLabels);
        int total = this.truePositives + this.trueNegatives + this.falsePositives + this.falseNegatives;
        if (total == 0) {
            return 0.0;
        }
        return (double) (this.truePositives + this.trueNegatives) / total;
    }

    public static void main(String[] args) {
        MetricsCalculator mc = new MetricsCalculator();
        int[] predictedLabels = {1, 1, 0, 0};
        int[] trueLabels = {1, 0, 0, 1};
        System.out.println(mc.precision(predictedLabels, trueLabels));
        System.out.println(mc.recall(predictedLabels, trueLabels));
        System.out.println(mc.f1Score(predictedLabels, trueLabels));
        System.out.println(mc.accuracy(predictedLabels, trueLabels));
    }
}

File: ClassEval_T/cpp/solution/MetricsCalculator.cpp
#include <vector>

class MetricsCalculator {
public:
    int true_positives;
    int false_positives;
    int false_negatives;
    int true_negatives;

    MetricsCalculator();

    void update(const std::vector<int>& predicted_labels, const std::vector<int>& true_labels);
    double precision(const std::vector<int>& predicted_labels, const std::vector<int>& true_labels);
    double recall(const std::vector<int>& predicted_labels, const std::vector<int>& true_labels);
    double f1_score(const std::vector<int>& predicted_labels, const std::vector<int>& true_labels);
    double accuracy(const std::vector<int>& predicted_labels, const std::vector<int>& true_labels);
};
MetricsCalculator::MetricsCalculator() {
    true_positives = 0;
    false_positives = 0;
    false_negatives = 0;
    true_negatives = 0;
}

void MetricsCalculator::update(const std::vector<int>& predicted_labels, const std::vector<int>& true_labels) {
    for (size_t i = 0; i < predicted_labels.size(); ++i) {
        int predicted = predicted_labels[i];
        int true_label = true_labels[i];

        if (predicted == 1 && true_label == 1) {
            true_positives++;
        }
        else if (predicted == 1 && true_label == 0) {
            false_positives++;
        }
        else if (predicted == 0 && true_label == 1) {
            false_negatives++;
        }
        else if (predicted == 0 && true_label == 0) {
            true_negatives++;
        }
    }
}

double MetricsCalculator::precision(const std::vector<int>& predicted_labels, const std::vector<int>& true_labels) {
    update(predicted_labels, true_labels);
    if (true_positives + false_positives == 0) {
        return 0.0;
    }
    return static_cast<double>(true_positives) / (true_positives + false_positives);
}

double MetricsCalculator::recall(const std::vector<int>& predicted_labels, const std::vector<int>& true_labels) {
    update(predicted_labels, true_labels);
    if (true_positives + false_negatives == 0) {
        return 0.0;
    }
    return static_cast<double>(true_positives) / (true_positives + false_negatives);
}

double MetricsCalculator::f1_score(const std::vector<int>& predicted_labels, const std::vector<int>& true_labels) {
    update(predicted_labels, true_labels);
    double precision_value = precision(predicted_labels, true_labels);
    double recall_value = recall(predicted_labels, true_labels);
    if (precision_value + recall_value == 0.0) {
        return 0.0;
    }
    return 2.0 * precision_value * recall_value / (precision_value + recall_value);
}

double MetricsCalculator::accuracy(const std::vector<int>& predicted_labels, const std::vector<int>& true_labels) {
    update(predicted_labels, true_labels);
    int total = true_positives + true_negatives + false_positives + false_negatives;
    if (total == 0) {
        return 0.0;
    }
    return static_cast<double>(true_positives + true_negatives) / total;
}
